### PR TITLE
Use 4096 bytes as default block size of the xfs and ext4 filesystem 

### PR DESCRIPTION
**Pull request by derekbit**
https://github.com/longhorn/longhorn/issues/4594

Signed-off-by: Derek Su <derek.su@suse.com>

### CODE REVIEW

**Comment by weizhe0422** [comment]
LGTM

It can support xfs and ext4 fs through different length of arguments.

**Comment by derekbit** [comment]
Ready for review.

cc @PhanLe1010 @shuo-wu @innobead

File: csi/node_server.go [hunk 1]
```diff
@@ -615,21 +628,27 @@ func (ns *NodeServer) getMounter(volume *longhornclient.Volume, volumeCapability
...
 		}
 
 		mounter := &mount.SafeFormatAndMount{Interface: mount.New(""), Exec: utilexec.New()}
-		// we allow the user to provide additional params for ext4 filesystem creation.
-		// this allows an ext4 fs to be mounted on older kernels, see https://github.com/longhorn/longhorn/issues/1208
-		if fsType == "ext4" && userExt4Params != nil && userExt4Params.Value != "" {
-			ext4Params := userExt4Params.Value
-			logrus.Infof("volume %v using user provided ext4 fs creation params: %s", volume.Name, ext4Params)
-			cmdParamMapping := map[string]string{"mkfs." + fsType: ext4Params}
+		if fsType == "ext4" || fsType == "xfs" {
```
Comment (shuo-wu) [inline csi/node_server.go:649]: Print out warning logs if the fs type is unrecognized.

**Review by shuo-wu — COMMENTED** [review]
In general LGTM

File: csi/node_server.go (hunk 1 above)
Comment (derekbit) [reply csi/node_server.go:649]: Updated, thank you.

File: csi/node_server.go [hunk 2]
```diff
@@ -615,25 +628,33 @@ func (ns *NodeServer) getMounter(volume *longhornclient.Volume, volumeCapability
...
+			fsType = DefaultFsType
+		}
+
+		// To allow users to override the default block size,
+		// put the default block size in front of other user-defined parameters.
+		params := getBlockSizeParams(fsType)
+
+		// We allow the user to provide additional params for ext4 filesystem creation.
+		// This allows an ext4 fs to be mounted on older kernels, see https://github.com/longhorn/longhorn/issues/1208
+		if fsType == "ext4" {
```
Comment (innobead) [inline csi/node_server.go:642]: Use `DefaultFsType` instead of `ext4`

File: csi/node_server.go [hunk 3]
```diff
@@ -615,25 +628,33 @@ func (ns *NodeServer) getMounter(volume *longhornclient.Volume, volumeCapability
...
 		}
 
 		mounter := &mount.SafeFormatAndMount{Interface: mount.New(""), Exec: utilexec.New()}
-		// we allow the user to provide additional params for ext4 filesystem creation.
-		// this allows an ext4 fs to be mounted on older kernels, see https://github.com/longhorn/longhorn/issues/1208
-		if fsType == "ext4" && userExt4Params != nil && userExt4Params.Value != "" {
-			ext4Params := userExt4Params.Value
-			logrus.Infof("volume %v using user provided ext4 fs creation params: %s", volume.Name, ext4Params)
-			cmdParamMapping := map[string]string{"mkfs." + fsType: ext4Params}
+		if fsType == "ext4" || fsType == "xfs" {
```
Comment (innobead) [inline csi/node_server.go:649]: The same here to use `DefaultFsType` or another two constants for ext4 and xfs.

**Review by shuo-wu — APPROVED** [review]

File: csi/node_server.go [hunk 4]
```diff
@@ -615,25 +628,33 @@ func (ns *NodeServer) getMounter(volume *longhornclient.Volume, volumeCapability
...
+			logrus.Infof("volume %v using user and longhorn provided %v fs creation params: %s", volume.Name, fsType, params)
+			cmdParamMapping := map[string]string{"mkfs." + fsType: params}
 			mounter = &mount.SafeFormatAndMount{
 				Interface: mount.New(""),
 				Exec:      NewForcedParamsExec(cmdParamMapping),
 			}
+		} else {
+			logrus.Infof("volume %v with unsupported filesystem %v, use default fs creation params", volume.Name, fsType)
```
Comment (innobead) [inline csi/node_server.go:657]: Should we use `Errorf` instead? also return an error if the fs not supported?

**Review by innobead — COMMENTED** [review]
LGTM, just minor comments.

File: csi/node_server.go (hunk 2 above)
Comment (derekbit) [reply csi/node_server.go]: Updated, thank you.

File: csi/node_server.go (hunk 3 above)
Comment (derekbit) [reply csi/node_server.go]: Updated, thank you.

File: csi/node_server.go (hunk 4 above)
Comment (derekbit) [reply csi/node_server.go]: Use `Warnf`, instead. Thank you.

**Review by innobead — APPROVED** [review]

**Comment by innobead** [comment]
@derekbit Please help backport to 1.3 and 1.2. Thanks.

**Review by joshimoo — COMMENTED** [review]
LGTM, good approach of reusing the forced params executor to extend the mkfs. call :)

**Comment by PhanLe1010** [comment]
LGTM for this release

However, this only fix ext4 and xfs. In the future, maybe we can allow the user to specify the filesystem parameter in storageclass for any filesystem. Ref: https://github.com/longhorn/longhorn/issues/4401#issuecomment-1212565935

How do you think @innobead @derekbit @joshimoo ?

**Comment by derekbit** [comment]
> LGTM for this release
> 
> However, this only fix ext4 and xfs. In the future, maybe we can allow the user to specify the filesystem parameter in storageclass for any filesystem. Ref: [longhorn/longhorn#4401 (comment)](https://github.com/longhorn/longhorn/issues/4401#issuecomment-1212565935)
> 
> How do you think @innobead @derekbit @joshimoo ?

Totally agreed! Actually, I was implementing this idea when working on this PR.
The idea is put the format (mkfs) options in `req.VolumeContext["formatOptions"]`.

In addition, we have ext4 mkfs parameters setting now. if users set the per-storage mkfs options, the options will override the settings from the global settings. How about?

**Comment by joshimoo** [comment]
Good idea to add the parameters to the storage class, we can deprecate the global setting in the 1.3.x release then remove it for the 1.4 release since the parameter is only relevant for new volume creation, no point in carry it around for a long time. Let's make sure it's part of the 1.3.x and 1.4 readmes since it's an important change any way where we would want the user to specify the 4k block size.

Just for reference the storage class params can also be used to address this issue https://github.com/longhorn/longhorn/issues/4129#issuecomment-1158466458

**Comment by innobead** [comment]
Sounds good to me, especially from 1.4, ideally, fs type should not be an issue anymore, so it's fair to make the fs format options general in the storage class instead.

**Comment by PhanLe1010** [comment]
Ceph CSI has some example: https://github.com/ceph/ceph-csi/blob/0311eb5f447762a581194d649ea4d703da104fb8/charts/ceph-csi-cephfs/templates/storageclass.yaml#L55

**Pull request merged**